Consolidate Metal3 Contributing guide (#40)
metal3-io/community: CONTRIBUTING.md

- [ironic-standalone-operator](https://github.com/metal3-io/ironic-standalone-operator/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22)

Alternatively, read the documentation, try to write your own code or
examples, and file issues for any problems or gaps in documentation that you
encounter!

## Contributing a Patch

1. If you haven't already done so, sign a Contributor License Agreement
   (see details in the [Kubernetes CLA documentation](https://github.com/kubernetes/community/blob/master/CLA.md)).
2. Fork the desired repository, develop and test your code changes.
3. Submit a pull request.

## Pull Request Labels

All code PRs should be labeled with one of the following to indicate the type of
change:

- ⚠️ (`:warning:`, major or breaking changes)
- ✨ (`:sparkles:`, feature additions)
- 🐛 (`:bug:`, patch and bugfixes)
- 📖 (`:book:`, documentation or proposals)
- 🌱 (`:seedling:`, minor or other)

Individual commits should not be tagged separately, but will generally be
assumed to match the PR. For instance, if you have a bugfix along with a
breaking change, it's generally encouraged to submit the bugfix separately,

> but if you must put them in one PR, mark the commits separately.

Reviewer: emojis don't go with commits, they go with PR only.

## Issue and Pull Request Management

Anyone may comment on issues and submit reviews for pull requests. However, in
order to be assigned an issue or pull request, you must be a member of the
[Metal3-io organization](https://github.com/metal3-io) on GitHub.

Metal3 maintainers can assign you an issue or pull request by leaving a
`/assign <your Github ID>` comment on the issue or pull request.

## Commands and Workflow

Metal3 projects follow the standard Kubernetes workflow and use Prow for CI/CD.

We use the same priority and kind labels as Kubernetes. See the labels tab in
GitHub for the full list in each repository.

The standard Kubernetes comment commands work in Metal3 projects. See
[Prow command help](https://prow.apps.test.metal3.io/command-help) for a
command reference.

Common commands include:

- `/lgtm` - Adds the `lgtm` label (Looks Good To Me)
- `/approve` - Adds the `approve` label
- `/hold` - Adds the `do-not-merge/hold` label
- `/hold cancel` - Removes the hold
- `/retest` - Reruns failed tests
- `/assign @username` - Assigns an issue or PR
- `/cc @username` - Requests a review

## Release Process

> ### General Release Guidelines

Reviewer: Main releasing.doc is unliked, while we do link the four repos individual release docs.

Metal3 projects follow these general release practices:

- **Minor and patch releases** are generally done immediately after a feature or

>   bugfix is landed, or sometimes a series of features tied together.

Reviewer 2: Isn't it more like monthly? We don't push releases as soon as there is anything merged to a release branch

- **Minor releases** will only be tagged on the *most recent* major release branch,
  except in exceptional circumstances. Patches will be backported to maintained
  stable versions, as needed.

> - **Major releases** are done shortly after a breaking change is merged. Once a

Reviewer 3: We do merge breaking changes in minor version.

- [Release Process](#release-process)
   - [Repo-Specific Release Steps](#repo-specific-release-steps)
- [Documentation Changes](#documentation-changes)
- [Proposal Process](#proposal-process)
- [Triaging Issues](#triaging-issues)
- [AI Agent Guidelines](#ai-agent-guidelines)
- [Google Doc Viewing Permissions](#google-doc-viewing-permissions)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Certificate of Origin

By contributing to this project you agree to the Developer Certificate of
Origin (DCO). This document was created by the Linux Kernel community and is a
simple statement that you, as a contributor, have the legal right to make the
contribution. See the [DCO](DCO) file for details.

### Git commit Sign-off

Commit message should contain signed off section with full name and email. For example:

```text
Signed-off-by: John Doe <jdoe@example.com>
```

When making commits, include the `-s` flag and `Signed-off-by` section will be
automatically added to your commit message. If you want GPG signing too, add
the `-S` flag alongside `-s`.

```bash
# Signing off commit
git commit -s

# Signing off commit and also additional signing with GPG

> git commit -s -S

Author: Agreed with Lennart and Tuomo, I left it as is for now

If you're new to the project and want to help, but don't know where to start,
look for issues labeled with `good first issue` and `help wanted` in the Metal3
repositories. Here's a consolidated link of good first issues and help wanted
across the Metal3.io org.

- [Good first issues](https://github.com/search?q=org%3Ametal3-io+label%3A%22good+first+issue%22+is%3Aissue+is%3Aopen&type=issues)
- [Help wanted](https://github.com/search?q=org%3Ametal3-io%20label%3A%22help%20wanted%22%20is%3Aissue%20is%3Aopen&type=issues)

Alternatively, read the documentation, try to write your own code or
examples, and file issues for any problems or gaps in documentation that you
encounter!

## Contributing a Patch

1. If working on an issue, signal other contributors that you are actively
   working on it by assigning it to yourself.
1. Fork the desired repo, develop and test your code changes.
1. Submit a pull request.
    1. All code PR must be labeled with one of
        - ⚠️ (`:warning:`, major or breaking changes)
        - ✨ (`:sparkles:`, feature additions)
        - 🐛 (`:bug:`, patch and bugfixes)
        - 📖 (`:book:`, documentation or proposals)
        - 🌱 (`:seedling:`, minor or other)

> 1. If your PR has multiple commits, you should

Reviewer 4: ```suggestion
1. If your PR has multiple commits related to the same change, you should
```

   [squash them into a single commit](https://kubernetes.io/docs/contribute/new-content/open-a-pr/#squashing-commits)
   before merging your PR unless it makes sense to have multiple commits.

Individual commits should not be tagged separately with emojis, but will
generally be assumed to match the PR. The emoji label should match the most
significant change in the PR. Test updates, refactoring, and other minor
changes should use 🌱 (`:seedling:`).

For instance, if you have a bugfix along with a breaking change, it's generally
encouraged to submit the bugfix separately, but if you must put them in one PR,
the PR should be labeled with the most significant change (⚠️ in this case).

All changes must be code reviewed. Coding conventions and standards are
explained in the official [developer docs](https://git.k8s.io/community/contributors/devel).

## Testing

> In Metal3.io, we generally run Kubernetes Prow tests and Jenkins tests in CI.

Reviewer: Make them actual links.

> Prow tests are usually run automatically and jenkins tests are triggered

Reviewer 4: ```suggestion
Prow tests usually run automatically and jenkins tests are triggered
```

manually. You can read more on the [Metal3.io prow configuration](https://github.com/metal3-io/project-infra/blob/main/prow/README.md)
and [Metal3.io jenkins setup](https://github.com/metal3-io/project-infra/blob/main/jenkins/README.md).

> All tests are expected to be pass before merging . There are quite a few handy

Reviewer 4: ```suggestion
All tests are expected to be pass before merging. There are quite a few handy
```

scripts and tools (linters, formatters etc) under the `/hack` folder in each

> repo. Please make use of those to run the tests locally before push.

Reviewer 4: ```suggestion
repo. Please make use of those to run the tests locally before pushing changes upstream.
```

  minor and patch releases are tagged. Branches are created usually when we do
  a release candidate (RC) release for a minor release.

- In some cases, it may be necessary to open PRs for bugfixes directly against
  stable branches, but this should generally not be the case.

### Release Branch Support

Each project maintains specific support matrices for their release branches.
Check [version support](https://book.metal3.io/version_support.html) for :

- Maintained versions and EOL dates
- API version support policies
- Test coverage policies (if any)

## Breaking Changes

Breaking changes are generally allowed in the `main` branch, as this is the
branch used to develop the next minor release. We encourage to merge such changes
in the beginning of a minor release cycle if possible so that the changes are
well tested.

There may be times, however, when `main` is closed for breaking changes. This
is likely to happen as we near the release of a new minor version.

Breaking changes are **not allowed** in release branches, as these represent
minor versions that have already been released. These versions have consumers

> who expect the APIs, behaviors, etc. to remain stable during the lifetime of

Reviewer 4: ```suggestion
who expect the APIs, operational logic, etc. to remain stable during the lifetime of
```

  Go minor version of a supported branch goes out of support (e.g., to pick up
  bug and CVE fixes)
- Improvements to existing documentation
- Improvements to the test framework and CI

### Backport Process

Like any other activity in the project, backporting a fix/change is a
community-driven effort and requires that someone volunteers to own the task.

In most cases, the cherry-pick bot can (and should) be used to automate
opening a cherry-pick PR. To use it, add a comment on the PR you want to
backport:

```text
/cherry-pick release-X.Y
```

### Maintained Versions

We generally do not accept backports to release branches that are not maintained
(including tested only branches).
Check the [Version Support](https://github.com/metal3-io/metal3-docs/blob/main/docs/user-guide/src/version_support.md)
guide for each project's specific support policy.

## Merge Approval

Please see the [Kubernetes community document on pull
requests](https://git.k8s.io/community/contributors/guide/pull-requests.md) for
more information about the merge process.

Metal3 follows the standard Kubernetes workflow: any PR needs `lgtm` and
`approved` labels, and PRs must pass all tests before being merged.

> 

Reviewer 4: Add please: Tests may be overwritten by maintainers in case the tests are not relevant for a particular PR or fail for unrelated CI issues.

the `-S` flag alongside `-s`.

```bash
# Signing off commit
git commit -s

# Signing off commit and also additional signing with GPG
git commit -s -S
```

## Finding Things That Need Help

If you're new to the project and want to help, but don't know where to start,
look for issues labeled with `good first issue` in any of the Metal3 repositories.
Here's a consolidated link of good first issues across the Metal3.io org.

- [Good first issues](https://github.com/search?q=org%3Ametal3-io%20label%3A%22help%20wanted%22%20is%3Aissue%20is%3Aopen&type=issues)

Alternatively, read the documentation, try to write your own code or
examples, and file issues for any problems or gaps in documentation that you
encounter!

## Contributing a Patch

1. If working on an issue, signal other contributors that you are actively
   working on it by assigning it to yourself.
1. Fork the desired repo, develop and test your code changes.
1. Submit a pull request.
    1. All code PR must be labeled with one of
        - ⚠️ (`:warning:`, major or breaking changes)
        - ✨ (`:sparkles:`, feature additions)
        - 🐛 (`:bug:`, patch and bugfixes)
        - 📖 (`:book:`, documentation or proposals)

>         - 🌱 (`:seedling:`, minor or other)

Reviewer: This should highlight that the test fixes etc should go into seedling, and not bug/feature/breaking.

> 1. If your PR has multiple commits, you must

Reviewer: I'm good with this version as it is not absolute anymore that you must squash :+1:

explained in the official [developer docs](https://git.k8s.io/community/contributors/devel).
Expect reviewers to request that you avoid common
[go style mistakes](https://github.com/golang/go/wiki/CodeReviewComments)
in your PRs.

## Testing

In Metal3.io, we generally run Kubernetes Prow tests and Jenkins tests in CI.
Prow tests are usually run automatically and jenkins tests are triggered
manually. You can read more on the [Metal3.io prow configuration](https://github.com/metal3-io/project-infra/blob/main/prow/README.md)
and [Metal3.io jenkins setup](https://github.com/metal3-io/project-infra/blob/main/jenkins/README.md).
All tests are expected to be pass before merging . There are quite a few handy
scripts and tools (linters, formatters etc) under the `/hack` folder in each
repo. Please make use of those to run the tests locally before push.

## Code Review Guidelines

All changes must be code reviewed. Coding conventions and standards are
explained in the official [Kubernetes developer
docs](https://github.com/kubernetes/community/tree/master/contributors/devel).

Expect reviewers to request that you avoid common [Go style

> mistakes](https://github.com/golang/go/wiki/CodeReviewComments) in your PRs.

Reviewer: This is duplicated from lines 107-109.

You will find more information on the review process in #merge-approval, #issue-and-pull-request-management,
and #commands-and-workflow section.

## Versioning and Release Semantics

### What Goes in a Release

Metal3 projects generally follow these versioning semantics:

- A (**minor**) release CAN include:
   - Introduction of new API versions, or new Kinds
   - Compatible API changes like field additions, deprecation notices, etc.
   - Breaking API changes for deprecated APIs, fields, or code
   - Features, promotion or removal of feature gates
   - And more!

A (**patch**) release SHOULD only include backwards compatible set of:

- bugfixes
- ci/testing fixes
- version bumps

## Branches

### Branch Structure

Metal3 repositories follow a common branch structure:

- The **main** branch is where development happens. All the latest and greatest
  code, including breaking changes, happens on main.

- The **release-X.Y** branches contain stable, backwards compatible code. On every
  major or minor release, a new branch is created. It is from these branches that
  minor and patch releases are tagged. Branches are created usually when we do

>   a release candidate (rc) release for a minor release.

Reviewer: ```suggestion
  minor and patch releases are tagged. Branches are created when we do
  a release candidate (RC) release for a minor release.
```

maintained branches:

- Critical bug fixes, security issue fixes, or fixes for bugs without easy workarounds
- Dependency bumps
- Changes required to support new Kubernetes patch versions, when possible
- Changes to use the latest Go patch version to build controller images
- Changes to bump the Go minor version used to build controller images, if the
  Go minor version of a supported branch goes out of support (e.g., to pick up
  bug and CVE fixes)
- Improvements to existing documentation
- Improvements to the test framework and CI

### Backport Process

Like any other activity in the project, backporting a fix/change is a
community-driven effort and requires that someone volunteers to own the task.

In most cases, the cherry-pick bot can (and should) be used to automate
opening a cherry-pick PR. To use it, add a comment on the PR you want to
backport:

```text
/cherry-pick release-X.Y
```

### Maintained Versions

We generally do not accept backports to release branches that are not maintained
(including tested only branches).
Check the [Version Support](https://github.com/metal3-io/metal3-docs/blob/main/docs/user-guide/src/version_support.md)

> guide for each project's specific support policy.

Reviewer: Especially when we have this section and link, we shouldn't list the stuff mentioned in L174 in contrib/readme.

or pull requests, as well as submission of official pull request reviews, is
limited to members of the [Metal3-io organization](https://github.com/metal3-io)
on GitHub.

Metal3 maintainers can assign you an issue or pull request by leaving a
`/assign <your Github ID>` comment on the issue or pull request.

## Commands and Workflow

Metal3 projects follow the standard Kubernetes workflow and use Prow for CI/CD.

We use the same priority and kind labels as Kubernetes. See the labels tab in
GitHub for the full list in each repository.

The standard Kubernetes comment commands work in Metal3 projects. See
[Prow command help](https://prow.apps.test.metal3.io/command-help) for a
command reference.

Common commands include:

- `/lgtm` - Adds the `lgtm` label (Looks Good To Me)
- `/lgtm cancel` - Removes the `lgtm` label
- `/approve` - Adds the `approve` label
- `/approve cancel` - Removes the `approve` label
- `/hold` - Adds the `do-not-merge/hold` label
- `/hold cancel` or `/unhold` - Removes the hold
- `/retest` - Reruns failed tests
- `/assign @username` - Assigns an issue or PR
- `/cc @username` - Requests a review
- `/test ?` - Lists all the test that one can run on a PR

> 

Reviewer: `/ok-to-test` is critical for getting tests run for external contributors

## Release Process

[Releasing in Metal3](https://github.com/metal3-io/metal3-docs/blob/main/processes/releasing.md)
contains the high-level release process for Metal3.io org and possible follow-up
actions.

### Repo-Specific Release Steps

Each repository has specific release procedures and automation. Refer to each
repository's documentation:

- [baremetal-operator releasing docs](https://github.com/metal3-io/baremetal-operator/blob/main/docs/releasing.md)
- [cluster-api-provider-metal3 releasing docs](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/docs/releasing.md)
- [ip-address-manager releasing docs](https://github.com/metal3-io/ip-address-manager/blob/main/docs/releasing.md)
- [ironic-standalone-operator releasing docs](https://github.com/metal3-io/ironic-standalone-operator/blob/main/docs/releasing.md)

> - Check individual project repos for their specific release documentation

Author: done
